lib/web_ui/test/canvaskit/fallback_fonts_test.dart

```diff
@@ -0,0 +1,127 @@
+  final LayerSceneBuilder sb = LayerSceneBuilder();
+  sb.pushOffset(0, 0);
+  sb.addPicture(ui.Offset.zero, picture);
+  dispatcher.rasterizer!.draw(sb.build().layerTree);
+  await matchGoldenFile(goldenFile,
+      region: region, maxDiffRatePercent: 0.0, write: write);
+}
+
+void testMain() {
+  setUpCanvasKitTest();
+
+  group('Font fallbacks', () {
+    /// Used to save and restore [ui.window.onPlatformMessage] after each test.
+    ui.PlatformMessageCallback? savedCallback;
+
+    setUp(() {
+      notoDownloadQueue.downloader = TestDownloader();
+      savedCallback = ui.window.onPlatformMessage;
+    });
+
+    tearDown(() {
+      ui.window.onPlatformMessage = savedCallback;
+    });
+
+    test('will download Noto Naskh Arabic if Arabic text is added', () async {
+      final Completer<void> fontChangeCompleter = Completer<void>();
+      // Intercept the system font change message.
+      ui.window.onPlatformMessage = (String name, ByteData? data,
+          ui.PlatformMessageResponseCallback? callback) {
+        if (name == 'flutter/system') {
+          const JSONMessageCodec codec = JSONMessageCodec();
+          final dynamic message = codec.decodeMessage(data);
+          if (message is Map) {
+            if (message['type'] == 'fontsChange') {
+              fontChangeCompleter.complete();
+            }
+          }
+        }
+        if (savedCallback != null) {
+          savedCallback!(name, data, callback);
+        }
+      };
+
+      // Creating this paragraph should cause us to start to download the
+      // fallback font.
+      CkParagraphBuilder pb = CkParagraphBuilder(
+        CkParagraphStyle(),
+      );
+      pb.addText('مرحبا');
+
+      await fontChangeCompleter.future;
+
+      expect(skiaFontCollection.globalFontFallbacks,
+          contains('Noto Naskh Arabic UI 0'));
+
+      final CkPictureRecorder recorder = CkPictureRecorder();
+      final CkCanvas canvas = recorder.beginRecording(kDefaultRegion);
+
+      pb = CkParagraphBuilder(
+        CkParagraphStyle(
+          fontSize: 32,
+        ),
+      );
+      pb.addText('مرحبا');
+      final CkParagraph paragraph = pb.build();
+      paragraph.layout(ui.ParagraphConstraints(width: 1000));
+
+      canvas.drawParagraph(paragraph, ui.Offset(200, 120));
+
+      await matchPictureGolden(
+        'canvaskit_font_fallback_arabic.png',
+        recorder.endRecording(),
+        write: true,
+      );
+    });
+  });
+}
+
+class TestDownloader extends NotoDownloader {
+  final Map<String, String> mockDownloads = <String, String>{
+    'https://fonts.googleapis.com/css2?family=Noto+Naskh+Arabic+UI': '''
+/* arabic */
+@font-face {
+  font-family: 'Noto Naskh Arabic UI';
+  font-style: normal;
+  font-weight: 400;
+  src: url(packages/ui/assets/NotoNaskhArabic-Regular.ttf) format('ttf');
+  unicode-range: U+0600-06FF, U+200C-200E, U+2010-2011, U+204F, U+2E41, U+FB50-FDFF, U+FE80-FEFC;
+}
+''',
```

Comment:
Let's also add a test where the server returns something illegible, just to make sure the app is not completely broken in that event.

Reply:
Done.